ibc-apps/ics721-nft-transfer/types/src/packet.rs

```diff
@@ -30,130 +23,42 @@ use crate::token::{TokenData, TokenIds, TokenUri};
 )]
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct PacketData {
+    #[cfg_attr(feature = "serde", serde(with = "serializers"))]
+    #[cfg_attr(feature = "schema", schemars(with = "String"))]
     pub class_id: PrefixedClassId,
     pub class_uri: Option<ClassUri>,
     pub class_data: Option<ClassData>,
     pub token_ids: TokenIds,
-    pub token_uris: Vec<TokenUri>,
-    pub token_data: Vec<TokenData>,
+    // Need `Option` to decode `null` value
+    pub token_uris: Option<Vec<TokenUri>>,
+    // Need `Option` to decode `null` value
+    pub token_data: Option<Vec<TokenData>>,
     pub sender: Signer,
     pub receiver: Signer,
-    pub memo: Memo,
+    pub memo: Option<Memo>,
 }
 
 impl PacketData {
-    #[allow(clippy::too_many_arguments)]
-    pub fn new(
-        class_id: PrefixedClassId,
-        class_uri: Option<ClassUri>,
-        class_data: Option<ClassData>,
-        token_ids: TokenIds,
-        token_uris: Vec<TokenUri>,
-        token_data: Vec<TokenData>,
-        sender: Signer,
-        receiver: Signer,
-        memo: Memo,
-    ) -> Result<Self, NftTransferError> {
-        if token_ids.0.is_empty() {
+    /// Performs the basic validation of the packet data fields.
+    pub fn validate_basic(&self) -> Result<(), NftTransferError> {
+        if self.token_ids.0.is_empty() {
             return Err(NftTransferError::NoTokenId);
         }
-        let num = token_ids.0.len();
-        let num_uri = token_uris.len();
-        let num_data = token_data.len();
+        let num = self.token_ids.0.len();
+        let num_uri = self
+            .token_uris
+            .as_ref()
+            .map(|t| t.len())
+            .unwrap_or_default();
+        let num_data = self
+            .token_data
+            .as_ref()
+            .map(|t| t.len())
+            .unwrap_or_default();
         if (num_uri != 0 && num_uri != num) || (num_data != 0 && num_data != num) {
             return Err(NftTransferError::TokenMismatched);
         }
-        Ok(Self {
-            class_id,
-            class_uri,
-            class_data,
-            token_ids,
-            token_uris,
-            token_data,
-            sender,
-            receiver,
-            memo,
-        })
-    }
-}
-
-impl TryFrom<RawPacketData> for PacketData {
```

Comment:
Let's keep these `TryFrom` and `From` conversions as they should work fine. IIUC, we hit the “null” issue only if someone tries to obtain PacketData proto types from JSON. It actually stems from the fact that the `.proto` definitions and hence the Go compiler allow nullable fields, which shouldn't be the case. I noticed that `ibc-go` handles this by adding `(gogoproto.nullable) = false` to their `.proto` files.

Actually, I am thinking of keeping the `nft-transfer` types' crate complete in terms of providing needed conversions, in case, any other implementors need it.